### PR TITLE
[Frontend] Feature/Notifications

### PR DESCRIPTION
**Pull request by 59023g**
Notifications will display `error`, `success` and `information` messages to the user. For example, it'll notify them when they're connected to testnet, when their Metamask connection failed, and when there's no liquidity

image

### CODE REVIEW

**Review by ryardley — CHANGES_REQUESTED** [review]
I think we want to set it up so that Notifications are only called from within usecases and it should look more like this:

Untitled Diagram (3)

File: ui/core/src/api/utils/NotificationService.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import {notifications} from "../../store/notifications"
+// i am pulling this from entitty, but id?
+// type NotificationServiceContext = {
+//   type: "error" | "success" | "inform"
+//   message: String
+// }
+
+// I dont think this will return anything? success, fail? boolean?
+type INotificationService = Boolean
```
[the remark this answers is not in the document]
Comment (ryardley) [reply ui/core/src/api/utils/NotificationService.ts:15]: Fire and forget we don't care about a return value here.

**Comment by 59023g** [comment]
> I think we want to set it up so that Notifications are only called from within usecases and it should look more like this:
> 

I agree, that would be good, if possible and applicable.

Two issues: 
1) Action service calls do not have return values. They use `updateData()` and `effect()` so I'm unable to set notifications in `actions.connectToWallet()` for example. We'd have to refactor the app to make this happen. 
2) In certain instances, for example `getWeb3Provider()`, this happens upon invocation of app via `const etheriumState = api.EthereumService.getState();` not by an action specifically.

File: ui/core/src/api/EthereumService/index.ts [hunk 2]
```diff
@@ -126,6 +134,7 @@ export class EthereumService implements IWalletService {
       }
 
       this.addWeb3Subscription();
+      notify({type: "success", message: "Connected to Metamask"})
```
[the remark this answers is not in the document]
Comment (59023g) [reply ui/core/src/api/EthereumService/index.ts:137]: See longer comment

File: ui/core/src/api/EthereumService/index.ts [hunk 3]
```diff
@@ -64,18 +65,25 @@ export class EthereumService implements IWalletService {
...
-        provider.on("disconnect", () => {
-          this.state.connected = false;
-        });
-      }
-      this.provider = provider;
-    });
+    getWeb3Provider() 
+      .then((provider) => {
+        if (!provider) {
+          notify({type: "error", message: "Metamask not found.", detail: "Check if extension enabled for this URL."})
```
[the remark this answers is not in the document]
Comment (59023g) [reply ui/core/src/api/EthereumService/index.ts:71]: I moved this to `utils/getMetamaskProvider()`

**Comment by 59023g** [comment]
> 1. Action service calls do not have return values. They use `updateData()` and `effect()` so I'm unable to set notifications in `actions.connectToWallet()` for example. We'd have to refactor the app to make this happen.

However, I realized I could remove `effect()`, invoke`connectToWallet()`, not have to refactor service return value, while still keeping the two states (ethService and store) in sync:

Screen Shot 2020-12-18 at 11 04 57 AM

**Comment by ryardley** [comment]
> Action service calls do not have return values. They use updateData() and effect() so I'm unable to set notifications in actions.connectToWallet() for example. We'd have to refactor the app to make this happen.

It can probably be done like this:

```ts
let firstRun = true;

effect(() => {
  if (!firstRun) {
    if(etheriumState.connected)
      notify({type: "success", message: "Connected to Metamask"})
    else 
      notify({type: "info", message: "Disconnected"})
  }   
  store.wallet.eth.isConnected = etheriumState.connected;
});

effect(() => {
  // ...
})

firstRun = false;
```

However this highlights the problems with using refs to share state outside of vue land. Semantically this is much better:

```ts
api.EthereumService.onConnected(() => {
  store.wallet.eth.isConnected = true
  notify({type: "success", message: "Connected!"})
})

api.EthereumService.onDisconnected(() => {
  store.wallet.eth.isConnected = false
  notify({type: "success", message: "Disconnected!"})
})
```

**Review by ryardley — APPROVED** [review]

**Pull request merged**